maas/client/bones/testing/server.py

```diff
@@ -27,12 +27,62 @@ def __init__(self, description):
         super(ApplicationBuilder, self).__init__()
         self._description = desc.Description(description)
         self._application = aiohttp.web.Application()
-        self._basepath, self._version = self._discover_version_and_base_path()
+        self._rootpath, self._basepath, self._version = (
+            self._discover_version_and_paths())
         self._wire_up_description()
         self._actions = {}
         self._views = {}
 
+    def route(self, method, path, handler=None):
+        """Add a handler for a specific path.
+
+        The path must start with a forward slash or be empty. The root URL of
+        the server is automatically discovered from the API description and is
+        added as a prefix to `path`, i.e. it's correct to specify "/accounts/"
+        but incorrect to specify "/MAAS/accounts/".
+
+        The handler is wrapped by `_wrap_handler` but should otherwise a normal
+        `aiohttp` request handler.
+
+        This method can be used as a decorator by omitting `handler`:
+
+          @builder.route("GET", "/foo/bar")
+          async def foobar(request):
+              ...
+
+        The use of `handle` should be preferred to `route` where the endpoint
+        is part of MAAS's Web API. This method allows mocking of miscellaneous
+        other paths when absolutely necessary.
+        """
```

Comment:
It's used in `test__obtains_credentials_from_server` and `test__raises_error_when_server_rejects_credentials` to set up handlers for `/accounts/authenticate/`, which is outside of MAAS's self-describing API.